### ISSUE TITLE
libpcre3 memory leak in Apache

### ISSUE THREAD

**Issue by prg3** [oerdnj/deb.sury.org#398]
Since upgrading to the libpcre3 package from the Sury repos, we have been seeing large memory usage in the apache instances. Downgrading to the debian release version for Trusty did correct this problem, however now this is no longer an option with the current dependencies on the packages.

Detection of the memory leak is by counting the (anon) memory blocks using pmap, and the upgraded nodes are 2-2.5x the number of anon blocks

for i in `ps aux | grep apache | awk '{print $2}'` ; do sudo pmap $i | grep anon | wc -l ; done | awk '{sum += $1} END {print sum}'

**Comment by oerdnj** [comment]
I have just uploaded pcre 8.39 to [ppa:ondrej/php-qa](https://launchpad.net/~ondrej/+archive/ubuntu/php-qa/+packages). Are you able to retry with this version? If it will not fix the bug, it might be worth reporting that to the upstream. Are you using any custom PHP code that's using PCRE and might be triggering some bug there?

**Comment by prg3** [comment]
I just uploaded this onto our staging environment and it appears to be functional, however the leak takes a while to become visible on our production systems, so this could take a bit to determine if we have resolved it.

There is custom code of course, however this is the same code that worked without any memory leaks prior to upgrading to the Sury repo version of libpcre3, so I suspect the issue is in the library and not the code.

Will update this ticket once we have been able to run through sufficient testing.

Thanks!

**Comment by oerdnj** [comment]
Oh, I am quite sure it's the library. The question is whether it's upstream bug or some Debian patch.

**Comment by prg3** [comment]
Very good question.. Anything else you need from me on this one? We are waiting and seeing what the behaviour of the latest update is but this will take some time (5-7 days) to get a good result.

**Comment by oerdnj** [comment]
If you can reproduce it with 8.39 it would be good to report it to upstream and have a test code that triggers it.

**Comment by oerdnj** [comment]
JFTR there was a security patch in Debian 8.38 that got a proper fix in upstream 8.39. The rest of Debian patches should have no effect on memory usage from what I recall.

**Comment by prg3** [comment]
Deployed with 8.39 approximately 12 hours ago, just waiting on the problem to either resurface or not now.

BTW, we are running Moodle as our primary application on this system.

**Comment by prg3** [comment]
This appears to still be an issue, memory use and anon blocks are increasing as expected. Unfortunately I have no idea what in the code base is triggering it, but I do know that we have identical code base systems running on Pcre 8.31 from the Ubuntu repositories that work just fine.

**Comment by oerdnj** [comment]
@prg3 It would be best if you could open a bug with upstream here: http://bugs.exim.org/buglist.cgi?product=PCRE&bug_status=NEW&bug_status=ASSIGNED&bug_status=REOPENED

**Comment by durzo** [comment]
I am also experiencing this issue, I temporarily downgraded to 7.0.7-4+deb.sury.org~trusty+1 but am now experiencing https://bugs.php.net/bug.php?id=72294 and need to upgrade php7.0-pgsql to version 7.0.8 to fix it.

is there any way you can compile against the libpcre that is bundled with ubuntu, instead of a newer upstream as it seems very broken at that end and could take a long time to fix even if it is reported ?

Otherwise I fear we will have to go back to php 5.6 :(

**Comment by oerdnj** [comment]
@durzo You can APT pin the libpcre3 version in trusty, see `man apt_preferences`. Something like:
```
cat << EOF > /etc/apt/preferences.d/libpcre3
Package: libpcre3
Pin: release n=trusty*
Pin-Priority: 990
EOF
```
should work.

**Comment by oerdnj** [comment]
Ah, I see. I updated the packages to depend only on `libpcre3 (>= 1:8.20)` instead of `libpcre3 (>= 2:8.20)` so it pull the new version on Ubuntu 12.04 where it's required. After the new PHP 7.0 packages finished the building, the pinning will worl.

**Comment by durzo** [comment]
Which build is that? I can see 7.0_7.0.8-4+deb.sury.org~trusty+1 but the buildlog shows it depends on libpcre3_2:8.39-1+deb.sury.org~trusty+1

**Comment by oerdnj** [comment]
@durzo Exactly, the hard dependency on `1:8.20` was introduced because anything less than that doesn't have JIT symbols required in PHP 7.0.

**Comment by durzo** [comment]
I have just deployed this to our prod cluster for testing and will report on the memory leak situation in a few days. thank you so much for your time

**Comment by prg3** [comment]
@oerdnj I can file a bug with them, however it's difficult because I have 0 idea how to reproduce this other than install and run it on the production cluster. I'm hoping using the new packages you released will help in my case as well.

**Comment by oerdnj** [comment]
The pcre3 is at 8.44 now
